x/oracle/types/tally.go

```diff
@@ -0,0 +1,173 @@
+package types
+
+import (
+	"container/heap"
+	"fmt"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type (
+	Vote interface {
+		codec.ProtoMarshaler
+
+		GetVoterAddress() string
+
+		GetVoteOption() VoteOption
+
+		GetConsensusValue() []byte
+	}
+
+	TallyKeeper interface {
+		Tally(sdk.Context, sdk.Iterator, Vote) (*TallyResult, error)
+	}
+)
+
+type Tally struct {
+	OracleValidatorInfos map[string]*OracleValidatorInfo
+	Yes                  map[string]*ConsensusTally
+	No                   sdk.Int
+	Total                sdk.Int
+}
+
+func NewTally() *Tally {
+	return &Tally{
+		OracleValidatorInfos: make(map[string]*OracleValidatorInfo),
+		Yes:                  make(map[string]*ConsensusTally),
+		No:                   sdk.ZeroInt(),
+		Total:                sdk.ZeroInt(),
+	}
+}
+
+// Add puts data to aggregate votes.
+func (t *Tally) Add(vote Vote) error {
+	oracleValidatorInfo, ok := t.OracleValidatorInfos[vote.GetVoterAddress()]
+	if !ok {
+		return fmt.Errorf("not found oracle. address: %s", vote.GetVoterAddress())
+	}
+	if !oracleValidatorInfo.IsPossibleVote() {
+		return nil
+	}
+
+	bondedTokens := oracleValidatorInfo.BondedTokens
+	switch vote.GetVoteOption() {
+	case VOTE_OPTION_VALID:
+		t.addYes(vote.GetConsensusValue(), bondedTokens)
+	case VOTE_OPTION_INVALID:
+		t.addNo(bondedTokens)
+	default:
+		return fmt.Errorf("unsupported voteOption. value: %s", vote.GetVoteOption())
+	}
+	return nil
+}
+
+// addYes defines to be divided and set by ConsensusValue
+func (t *Tally) addYes(key []byte, amount sdk.Int) {
+	if val, ok := t.Yes[string(key)]; ok {
+		val.VotingAmount = val.VotingAmount.Add(amount)
+	} else {
+		t.Yes[string(key)] = &ConsensusTally{
+			ConsensusKey: key,
+			VotingAmount: amount,
+		}
+	}
+}
+
+func (t *Tally) addNo(amount sdk.Int) {
+	t.No = t.No.Add(amount)
+}
+
+func (t *Tally) addTotal(amount sdk.Int) {
+	t.Total = t.Total.Add(amount)
+}
+
+// calculateTotal calculates the total share based on the registered OracleValidatorInfo.
+func (t *Tally) calculateTotal() {
+	for _, info := range t.OracleValidatorInfos {
+		t.addTotal(info.BondedTokens)
+	}
+}
+
+// CalculateTallyResult calculates the voting result based on the received quorum, creates and returns a TallyResult.
+func (t Tally) CalculateTallyResult(quorum sdk.Dec) *TallyResult {
+	t.calculateTotal()
+
+	tallyHeap := NewConsensusTallyMaxHeap()
+	for _, tally := range t.Yes {
+		heap.Push(&tallyHeap, tally)
+	}
+
+	tallyResult := NewTallyResult()
+
+	if tallyHeap.Len() > 0 {
+		maxTally := heap.Pop(&tallyHeap).(*ConsensusTally)
+		tallyResult.Yes = maxTally.VotingAmount
+
+		voteRate := maxTally.VotingAmount.ToDec().Quo(t.Total.ToDec())
+		if voteRate.GTE(quorum) {
+			tallyResult.ConsensusValue = maxTally.ConsensusKey
+		}
```

Comment:
If `voteRate` is less than `quorum`, I think calculating tally should be pended to wait more votes from other oracles, or decide it as invalid tally and remove it.

Reply:
I thought that once the voting was over, the voters could no longer vote.
I'm going to put the UpTime feature for participating in the missing vote here. :)

Reply:
Oops. 
Come to think of it, quorum seems to be a value that includes all of Yes, No, and InvalidYes, not only Yes.
I will correct this part.

Reply:
And if there are a lot of No votes, I don't think we should put the value of consensusValue.

Reply:
> I thought that once the voting was over, the voters could no longer vote.

Right. 

```go
for tallyHeap.Len() > 0 {
			invalidYesTally := heap.Pop(&tallyHeap).(*ConsensusTally)
			tallyResult.InvalidYes = tallyResult.InvalidYes.Add(invalidYesTally.VotingAmount)
		}

tallyResult.No = t.No

tallyResult.Total = t.Total
```

I think the above code block need to be moved inside the `if voteRate.GTE(quorum)` statement. And if max yes is less than quorum, this vote should be invalid and add additional logic for handling the invalid vote (such as updating status of OracleRegistration).

Reply:
Thank you! I modifed it as you said. :)

Reply:
Thanks @gyuguen!

And I think `threshold` is more proper name than `quorum`. Because quorum means 'the smallest number of people who must be at a meeting before it can begin or decisions can be made'. (https://www.oxfordlearnersdictionaries.com/definition/american_english/quorum)

Reply:
Oh I understand. I think it would be good to change the name.

Reply:
I fixed it. :)